esmvaltool/diag_scripts/weighting/climwip/core_functions.py

```diff
@@ -162,19 +196,52 @@ def calculate_weights(
     """
     if performance is not None:
         numerator = np.exp(-((performance / performance_sigma)**2))
+        # nans in the performance vector indicate models to be excluded
+        not_nan = np.isfinite(performance)
     else:
         numerator = 1
+        not_nan = True
     if independence is not None:
-        exp = np.exp(-((independence / independence_sigma)**2))
+        # don't consider nan models for independence of other models!
+        exp = np.exp(-((independence[:, not_nan] / independence_sigma)**2))
         # Note diagonal = exp(0) = 1, thus this is equal to 1 + sum(i!=j)
-        denominator = exp.sum('perfect_model_ensemble')
+        denominator = exp.sum(axis=1)
     else:
         denominator = 1
 
     weights = numerator / denominator
+    weights /= weights.sum(where=not_nan)
+
+    return weights
+
 
-    # Normalize weights
-    weights /= weights.sum()
+def calculate_weights(
+        performance: Union['xr.DataArray',
+                           None], independence: Union['xr.DataArray', None],
+        performance_sigma: Union[float, None],
+        independence_sigma: Union[float, None]) -> 'xr.DataArray':
+    """Xarray wrapper for calculate_weights_data."""
+    if performance is None:
+        performance_core_dims = []
+    else:
+        performance_core_dims = ['model_ensemble']
+    if independence is None:
+        independence_core_dims = []
+    else:
+        independence_core_dims = ['model_ensemble', 'model_ensemble_reference']
```

Comment:
You could also do this, to do it in a more 'pythonic' way:

```suggestion
    performance_core_dims = [] if performance is None else ['model_ensemble']
    independence_core_dims = [] if independence is None else ['model_ensemble', 'model_ensemble_reference']
```